### PR TITLE
feat: Refactor convex hull algorithm and improve performance

### PR DESCRIPTION
**Pull request by SamuelBehrmann**

### CODE REVIEW

File: src/convex-hull.ts [hunk 1]
```diff
@@ -1,71 +1,136 @@
...
+  let nextPoint;
+  while (!nextPoint?.equals(pointsOfRawConvexHull[0])) {
+    const neighbors = points.filter(
+      (point) =>
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 2].equals(
+          point
+        ) &&
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 1].equals(point)
+    );
+    if (!neighbors) {
```
Comment (luis-stumpf) [inline src/convex-hull.ts]: you can not check for empty like this you have to use the length because filter always returns at least an empty array

**Review by luis-stumpf — CHANGES_REQUESTED** [review]

File: src/convex-hull.ts [hunk 2]
```diff
@@ -1,71 +1,136 @@
...
+  // Find next neighbour to ensure the order of the points / lines is correct
+  let next = pointsResult[1];
+  while (!next?.equals(pointsResult[0])) {
+    next = adjacencyMatrix
+      .getNeighbors(pointsResult[pointsResult.length - 1])
+      .find(
+        (point) =>
+          !pointsResult[pointsResult.length - 2].equals(point) &&
+          !pointsResult[pointsResult.length - 1].equals(point)
+      )!;
```
Comment (luis-stumpf) [inline src/convex-hull.ts:127]: need do break if next is start point

File: src/convex-hull.ts [hunk 3]
```diff
@@ -1,71 +1,136 @@
...
+    if (!neighbors) {
+      break;
+    }
+
+    nextPoint = findPointWithBiggestClockwiseAngle(
+      pointsOfRawConvexHull[pointsOfRawConvexHull.length - 2],
+      pointsOfRawConvexHull[pointsOfRawConvexHull.length - 1],
+      neighbors
+    );
+
```
Comment (luis-stumpf) [inline src/convex-hull.ts:52]: break if nextpoint is start point otherwise we push it 2 times in the result array

File: src/convex-hull.ts (hunk 1 above)
Comment (SamuelBehrmann) [reply src/convex-hull.ts]: Done

File: src/convex-hull.ts [hunk 4]
```diff
@@ -1,71 +1,136 @@
...
+  let nextPoint;
+  while (!nextPoint?.equals(pointsOfRawConvexHull[0])) {
+    const neighbors = points.filter(
+      (point) =>
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 2].equals(
+          point
+        ) &&
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 1].equals(point)
+    );
+    if (!neighbors || neighbors?.length <= 0) {
```
Comment (luis-stumpf) [inline src/convex-hull.ts]: remove the !neighbors check this will always be false. because an empty array ist truthy in js

**Review by luis-stumpf — CHANGES_REQUESTED** [review]

File: src/convex-hull.ts (hunk 4 above)
Comment (SamuelBehrmann) [reply src/convex-hull.ts]: Can neighbours nicht undefined sein ? :thinking:

File: src/convex-hull.ts [hunk 5]
```diff
@@ -1,71 +1,143 @@
...
+  let nextPoint;
+  while (!nextPoint?.equals(pointsOfRawConvexHull[0])) {
+    const neighbors = points.filter(
+      (point) =>
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 2].equals(
+          point
+        ) &&
+        !pointsOfRawConvexHull[pointsOfRawConvexHull.length - 1].equals(point)
+    );
+    if (neighbors?.length <= 0) {
```
Comment (luis-stumpf) [inline src/convex-hull.ts:43]:
```suggestion
    if (neighbors.length === 0) {
```

**Review by luis-stumpf — APPROVED** [review]

**Pull request merged**